### PR TITLE
[GEOS-11443] REST API does not take effect immediately due to 10 minute authentication cache

### PR DESCRIPTION
**Pull request by petersmythe**
[GEOS-11443](https://osgeo-org.atlassian.net/browse/GEOS-11443) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

From my experience with the GeoServer REST API, e.g. adding a user to a group [GeoServer API Docs](https://docs.geoserver.org/stable/en/api/#1.0.0/usergroup.yaml) , the change happens and is written to disk immediately, however there is a 10 minute authentication cache that prevents the change from being **effective** immediately.

If I am impatient, I typically reload the configuration or use the AuthKey synchronise button to clear the 10 min auth cache. I am aware of the reload API endpoint [GeoServer API Docs](https://docs.geoserver.org/stable/en/api/#1.0.0/reload.yaml)  but that is a bit drastic.  
The reset endpoint does not appear to clear the auth cache.  This now fixes that.

A better solution than my previous attempt: https://github.com/geoserver/geoserver/pull/7746

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by petersmythe** [comment]
https://docs.geoserver.org/latest/en/user/rest/api/reset.html: 

> Resets all store, raster, and schema caches. This operation is used to **force GeoServer to drop all caches** and store connections and reconnect to each of them the next time they are needed by a request. This is useful in case the stores themselves cache some information about the data structures they manage that may have changed in the meantime.

The documentation is still perfectly applicable, no changes needed, unless we add `Resets all authentication, store, raster, and schema caches`?

File: src/restconfig/src/main/java/org/geoserver/rest/CatalogReloadController.java [hunk 1]
```diff
@@ -33,5 +35,8 @@ public void reload() throws Exception {
             method = {RequestMethod.POST, RequestMethod.PUT})
     public void reset() {
         geoServer.reset();
+        GeoServerExtensions.bean(GeoServerSecurityManager.class)
+                .getAuthenticationCache()
+                .removeAll();
```
Comment (aaime) [inline src/restconfig/src/main/java/org/geoserver/rest/CatalogReloadController.java:40]: This looks much better indeed, but... it's still backwards.
See how the controller is ust calling the reset method, without any explicit, special action targeting a particular cache? E.g., it's not calling ResourcePool to reset its caches, it's the pool that' listening to the reset events and cleaning itself. I believe you can have the security manager implement GeoServerLifecycleHandler and react on the reset "onReset" event instead. Or have you already tried it, and it did not work?

Comment (petersmythe) [reply src/restconfig/src/main/java/org/geoserver/rest/CatalogReloadController.java]: Implemented, successfully too, as you suggested (I think), but I am getting test errors in AuthenticationCacheFilterTest.java: 

> Caused by: org.geoserver.platform.GeoServerExtensions$MultipleBeansException: Multiple beans of type [org.geoserver.security](http://org.geoserver.security/).auth.AuthenticationCache


Clearly I still do not know how to instantiate the bean correctly.

**Comment by petersmythe** [comment]
@jodygarnett  Here's a summary of the changes, excluding documentation, etc.

`src/main/src/main/java/applicationContext.xml`: create a org.geoserver.security.auth.GuavaAuthenticationCacheImpl, 

why is it necessary to do this, given  https://github.com/geoserver/geoserver/blob/74ddad7db2179de4c9cc1cd266b73dceb938f98a/src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java#L698

```
    AuthenticationCache lookupAuthenticationCache() {
        AuthenticationCache authCache = GeoServerExtensions.bean(AuthenticationCache.class);
        return authCache != null ? authCache : new GuavaAuthenticationCacheImpl(1000);
    }
```
	

`src/main/src/main/java/org/geoserver/security/auth/GuavaAuthenticationCacheImpl.java` 	extend  GuavaAuthenticationCacheImpl to also  implements GeoServerLifecycleHandler (onReset calls removeAll)

The rest of the changes are superfluous

**Comment by petersmythe** [comment]
You ask:

> Are there any examples of how to select an alternate cache? 

I don't know if the code comment above helps?

> Or do you have to fix the original …

I'm not sure what you mean here, the original what?  `GuavaAuthenticationCacheImpl` ?

**Comment by petersmythe** [comment]
> https://github.com/geoserver/geoserver/blob/79239c73bc747c3ed529ad0d2e6ee5ea02bc7e60/src/main/src/main/java/applicationContext.xml#L361

If I do not add this bean definition here, and leave the `GuavaAuthenticationCacheImpl` creation to https://github.com/geoserver/geoserver/blob/79239c73bc747c3ed529ad0d2e6ee5ea02bc7e60/src/main/src/main/java/org/geoserver/security/auth/GuavaAuthenticationCacheImpl.java#L31

then I do not see the GuavaAuthenticationCacheImpl listed in `GeoServerLifecycleHandler`s:

image

(and the required fix does not work)

**Pull request merged**